pool.go

```diff
@@ -334,13 +349,44 @@ func (p *Pool) err(err error) {
 	}
 }
 
+func (p *Pool) traceConnectDone(conn *ioErrConn, reason string, err error) {
+	if p.opts.pt.ConnectDone != nil {
+		var connectTime time.Duration
+		if conn != nil {
+			connectTime = conn.ConnectTime
+		} else {
+			connectTime = 0
+		}
+
+		p.opts.pt.ConnectDone(trace.PoolConnectDone{
+			PoolInfo:     trace.PoolInfo{Addr: p.addr},
+			PoolConnInfo: trace.PoolConnInfo{PoolSize: p.size, BufferSize: p.opts.overflowSize, AvailCount: len(p.pool)},
+			Reason:       reason,
+			ConnectTime:  connectTime,
+			Err:          err,
+		})
+	}
+}
+
+func (p *Pool) traceConnClosed(reason string) {
+	if p.opts.pt.ConnClosed != nil {
+		p.opts.pt.ConnClosed(trace.PoolConnClosed{
+			PoolInfo:     trace.PoolInfo{Addr: p.addr},
+			PoolConnInfo: trace.PoolConnInfo{PoolSize: p.size, BufferSize: p.opts.overflowSize, AvailCount: len(p.pool)},
+			Reason:       reason,
+		})
+	}
+}
+
 // this must always be called with p.l unlocked
 func (p *Pool) newConn(errIfFull bool) (*ioErrConn, error) {
+	start := time.Now()
 	c, err := p.opts.cf(p.network, p.addr)
+	elapsed := time.Since(start)
 	if err != nil {
 		return nil, err
 	}
-	ioc := newIOErrConn(c)
+	ioc := newIOErrConn(c, elapsed)
```

Comment:
Instead of tracking the time inside the `*ioErrConn` and tracing the event in the caller, why not just trace the event here? If it's for the `reason` phrase, you could just pass it as an argument to `newConn` (which I can also imagine being nice when debugging).

Also that way we could trace the time it took (could be interesting for timeouts).

Reply:
Yes, that was the reason. I'll change it not to hold tracked time in newIOErrConn and make newConn function handle reason.